src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,983 @@
+              node
+
+          node.cells(field)._pointee = Some(Slice(isGlobal(address).get._2._1.cells(0), 0))
+          address = relocatedAddress
+      }
+  )
+
+  externalFunctions.foreach(
+    external =>
+      val node = DSN(Some(this))
+      node.allocationRegions.add(DataLocation(external.name, external.offset, 0))
+      node.flags.global = true
+      node.flags.incomplete = true
+      globalMapping.update(AddressRange(external.offset, external.offset), Field(node, 0))
+  )
+
+
+
+  // determine if an address is a global and return the corresponding global if it is.
+  def isGlobal(address: BigInt): Option[(AddressRange, Field)] =
+    var global: Option[(AddressRange, Field)]  = None
+    breakable {
+      for (elem <- globalMapping) {
+        val range = elem._1 // TODO
+        if address >= range.start && (address < range.end || (range.start == range.end && range.end == address)) then
+          global = Some(elem)
+          break
+      }
+    }
+    global
+
+  def getCells(pos: CFGPosition, arg: Variable): Set[Slice] =
+    if reachingDefs(pos).contains(arg) then
+      reachingDefs(pos)(arg).foldLeft(Set[Slice]()) {
+        (s, defintion) =>
+          s + varToCell(defintion)(arg)
+      }
+    else
+      Set(formals(arg))
+
+
+  /**
+   * collects all the nodes that are currently in the DSG and updates nodes member variable
+   */
+  def collectNodes =
+    nodes.clear()
+    nodes.addAll(formals.values.map(_._1.node.get).map(n => find(n).node))
+    varToCell.values.foreach(
+      value => nodes.addAll(value.values.map(_._1.node.get).map(n => find(n).node))
+    )
+    nodes.addAll(stackMapping.values.map(n => find(n).node))
+    nodes.addAll(globalMapping.values.map(_._1).map(n => find(n).node))
+
+    val queue: mutable.Queue[DSN] = mutable.Queue()
+    queue.enqueueAll(nodes)
+    while queue.nonEmpty do
+      val cur = queue.dequeue()
+      cur.cells.foreach {
+        case (offset: BigInt, cell: DSC) if cell._pointee.isDefined =>
+          val node = find(cell.getPointee.node).node
+          if !nodes.contains(node) then
+            nodes.add(node)
+            queue.enqueue(node)
+        case _ =>
+      }
+
+  def toDot: String = {
+    collectNodes
+
+
+    var structs = nodes.foldLeft(Set[DotStruct]()) {
+      (s, n) =>
+        s + DotStruct(n.id.toString, n.toString, Some(n.cells.keys.map(o => o.toString)))
+    }
+
+    structs ++= formals.foldLeft(Set[DotStruct]()) {
+      (s, n) =>
+        val variable = n._1.name
+        s + DotStruct(s"Formal_$variable", s"Formal_$variable", None)
+    }
+
+    structs ++= varToCell.foldLeft(Set[DotStruct]()) {
+      (s, r) =>
+        val pos = r._1
+        val mapping = r._2
+        s ++ mapping.foldLeft(Set[DotStruct]()) {
+          (k, n) =>
+            val variable = n._1.name
+            k + DotStruct(s"SSA_${pos.toShortString.slice(1, 9)}_$variable", s"SSA_${pos}_$variable", None, false)
```

Comment:
This cuts off the first character of `pos.toShortString` which can cause problems for readability.